{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9c0733d4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/miniconda3/envs/ethical_necromancy/lib/python3.11/site-packages/Bio/pairwise2.py:278: BiopythonDeprecationWarning: Bio.pairwise2 has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.PairwiseAligner as a replacement, and contact the Biopython developers if you still need the Bio.pairwise2 module.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "\n",
    "def calculate_alignment_score(seq1, seq2):\n",
    "    alignments = pairwise2.align.globalxx(seq1, seq2)\n",
    "    # Use the best alignment\n",
    "    best = alignments[0]\n",
    "    identity = best[2] / best[4] * 100  # matches / aligned length * 100\n",
    "    return identity\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d9f8b26e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>May_2025_Strain_Locus</th>\n",
       "      <th>May_2025_Nucleotide_Seq</th>\n",
       "      <th>Published_KRhaeticus_Locus</th>\n",
       "      <th>Published_KRhaeticus_Seq</th>\n",
       "      <th>Jan_2025_Strain_Locus</th>\n",
       "      <th>Jan_2025_Nucleotide_Seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0:24410-25264:CENAFE_00150 -</td>\n",
       "      <td>tcaggaaggttttttccctgtgcagtttcctccatccggccaaccg...</td>\n",
       "      <td>1:3491885-3492739:GWK63_16165 +</td>\n",
       "      <td>tcaggaaggttttttccctgtgcagtttcctccatccggccaaccg...</td>\n",
       "      <td>2:3804957-3805811:EMBJFE_03523 +</td>\n",
       "      <td>tcaggaaggttttttccctgtgcagtttcctccatccggccaaccg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0:25274-25978:CENAFE_00155 -</td>\n",
       "      <td>tcacgtctcatcattagcgacatggtggttttggcggacaatctgc...</td>\n",
       "      <td>1:3491171-3491875:GWK63_16160 +</td>\n",
       "      <td>tcacgtctcatcattagcgacatggtggttttggcggacaatctgc...</td>\n",
       "      <td>2:3804243-3804947:EMBJFE_03522 +</td>\n",
       "      <td>tcacgtctcatcattagcgacatggtggttttggcggacaatctgc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0:25975-26670:CENAFE_00160 -</td>\n",
       "      <td>tcatgctcttgcctccacatgatccttgcctgaatagccagtacac...</td>\n",
       "      <td>1:3490479-3491174:GWK63_16155 +</td>\n",
       "      <td>tcatgctcttgcctccacatgatccttgcctgaatagccagtacac...</td>\n",
       "      <td>2:3803551-3804246:EMBJFE_03521 +</td>\n",
       "      <td>tcatgctcttgcctccacatgatccttgcctgaatagccagtacac...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0:26673-28205:CENAFE_00165 -</td>\n",
       "      <td>ttatccccctttcaccgtatgacgacgaatttccggtcttacaccc...</td>\n",
       "      <td>1:3488944-3490476:GWK63_16150 +</td>\n",
       "      <td>ttatccccctttcaccgtatgacgacgaatttccggtcttacaccc...</td>\n",
       "      <td>2:3802016-3803548:EMBJFE_03520 +</td>\n",
       "      <td>ttatccccctttcaccgtatgacgacgaatttccggtcttacaccc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0:28258-29793:CENAFE_00170 -</td>\n",
       "      <td>tctcactatgtggggcaggaaaagctacgcccatccgcagggtggg...</td>\n",
       "      <td>1:3486897-3488891:GWK63_16145 +</td>\n",
       "      <td>tctcactatgtggggcaggaaaagctacgcccatccgcagggtggg...</td>\n",
       "      <td>2:3800428-3801963:EMBJFE_03519 +</td>\n",
       "      <td>tctcactatgtggggcaggaaaagctacgcccatccgcagggtggg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3375</th>\n",
       "      <td>0</td>\n",
       "      <td>----------------------------------------------...</td>\n",
       "      <td>1:3267425-3267895:GWK63_15000 +</td>\n",
       "      <td>atgagtcagtcagtcaatcgcgccatcgtcatcggtcacctgggcc...</td>\n",
       "      <td>2:3896039-3896509:EMBJFE_03611 +</td>\n",
       "      <td>atgagtcagtcagtcaatcgcgccatcgtcatcggtcacctgggcc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3376</th>\n",
       "      <td>0</td>\n",
       "      <td>atgtgcgtcaaatcagacaacacgatccacgcgaattttacggacg...</td>\n",
       "      <td>1:3282706-3282927:GWK63_15085 +</td>\n",
       "      <td>atgtgcgtcaaatcagacaacacgatccacgcgaattttacggacg...</td>\n",
       "      <td>2:3701650-3701877:EMBJFE_03421 +</td>\n",
       "      <td>atgtgcgtcaaatcagacaacacgatccacgcgaattttacggacg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3377</th>\n",
       "      <td>0</td>\n",
       "      <td>----------------------------------------------...</td>\n",
       "      <td>1:3329074-3330267:GWK63_15335 -</td>\n",
       "      <td>ttatgatgatgggtctggcagtgccggttcttctccgaaggggcga...</td>\n",
       "      <td>2:3627266-3628459:EMBJFE_03325 +</td>\n",
       "      <td>ttatgatgatgggtctggcagtgccggttcttccccgaaggggcga...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3378</th>\n",
       "      <td>0</td>\n",
       "      <td>tc--ttacgctgagaaaattctcattcaaaagtca-ctctttttat...</td>\n",
       "      <td>1:3386823-3387596:GWK63_15655 -</td>\n",
       "      <td>tcattttcggtcagaca---ctaacgcatgagccagcatgtttcgc...</td>\n",
       "      <td>2:1463094-1463867:EMBJFE_01316 -</td>\n",
       "      <td>tcattttcggtcagaca---ctaacgcatgagccagcatgtttcgc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3379</th>\n",
       "      <td>\"</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3380 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             May_2025_Strain_Locus  \\\n",
       "0     0:24410-25264:CENAFE_00150 -   \n",
       "1     0:25274-25978:CENAFE_00155 -   \n",
       "2     0:25975-26670:CENAFE_00160 -   \n",
       "3     0:26673-28205:CENAFE_00165 -   \n",
       "4     0:28258-29793:CENAFE_00170 -   \n",
       "...                            ...   \n",
       "3375                             0   \n",
       "3376                             0   \n",
       "3377                             0   \n",
       "3378                             0   \n",
       "3379                             \"   \n",
       "\n",
       "                                May_2025_Nucleotide_Seq  \\\n",
       "0     tcaggaaggttttttccctgtgcagtttcctccatccggccaaccg...   \n",
       "1     tcacgtctcatcattagcgacatggtggttttggcggacaatctgc...   \n",
       "2     tcatgctcttgcctccacatgatccttgcctgaatagccagtacac...   \n",
       "3     ttatccccctttcaccgtatgacgacgaatttccggtcttacaccc...   \n",
       "4     tctcactatgtggggcaggaaaagctacgcccatccgcagggtggg...   \n",
       "...                                                 ...   \n",
       "3375  ----------------------------------------------...   \n",
       "3376  atgtgcgtcaaatcagacaacacgatccacgcgaattttacggacg...   \n",
       "3377  ----------------------------------------------...   \n",
       "3378  tc--ttacgctgagaaaattctcattcaaaagtca-ctctttttat...   \n",
       "3379                                               None   \n",
       "\n",
       "           Published_KRhaeticus_Locus  \\\n",
       "0     1:3491885-3492739:GWK63_16165 +   \n",
       "1     1:3491171-3491875:GWK63_16160 +   \n",
       "2     1:3490479-3491174:GWK63_16155 +   \n",
       "3     1:3488944-3490476:GWK63_16150 +   \n",
       "4     1:3486897-3488891:GWK63_16145 +   \n",
       "...                               ...   \n",
       "3375  1:3267425-3267895:GWK63_15000 +   \n",
       "3376  1:3282706-3282927:GWK63_15085 +   \n",
       "3377  1:3329074-3330267:GWK63_15335 -   \n",
       "3378  1:3386823-3387596:GWK63_15655 -   \n",
       "3379                             None   \n",
       "\n",
       "                               Published_KRhaeticus_Seq  \\\n",
       "0     tcaggaaggttttttccctgtgcagtttcctccatccggccaaccg...   \n",
       "1     tcacgtctcatcattagcgacatggtggttttggcggacaatctgc...   \n",
       "2     tcatgctcttgcctccacatgatccttgcctgaatagccagtacac...   \n",
       "3     ttatccccctttcaccgtatgacgacgaatttccggtcttacaccc...   \n",
       "4     tctcactatgtggggcaggaaaagctacgcccatccgcagggtggg...   \n",
       "...                                                 ...   \n",
       "3375  atgagtcagtcagtcaatcgcgccatcgtcatcggtcacctgggcc...   \n",
       "3376  atgtgcgtcaaatcagacaacacgatccacgcgaattttacggacg...   \n",
       "3377  ttatgatgatgggtctggcagtgccggttcttctccgaaggggcga...   \n",
       "3378  tcattttcggtcagaca---ctaacgcatgagccagcatgtttcgc...   \n",
       "3379                                               None   \n",
       "\n",
       "                 Jan_2025_Strain_Locus  \\\n",
       "0     2:3804957-3805811:EMBJFE_03523 +   \n",
       "1     2:3804243-3804947:EMBJFE_03522 +   \n",
       "2     2:3803551-3804246:EMBJFE_03521 +   \n",
       "3     2:3802016-3803548:EMBJFE_03520 +   \n",
       "4     2:3800428-3801963:EMBJFE_03519 +   \n",
       "...                                ...   \n",
       "3375  2:3896039-3896509:EMBJFE_03611 +   \n",
       "3376  2:3701650-3701877:EMBJFE_03421 +   \n",
       "3377  2:3627266-3628459:EMBJFE_03325 +   \n",
       "3378  2:1463094-1463867:EMBJFE_01316 -   \n",
       "3379                              None   \n",
       "\n",
       "                                Jan_2025_Nucleotide_Seq  \n",
       "0     tcaggaaggttttttccctgtgcagtttcctccatccggccaaccg...  \n",
       "1     tcacgtctcatcattagcgacatggtggttttggcggacaatctgc...  \n",
       "2     tcatgctcttgcctccacatgatccttgcctgaatagccagtacac...  \n",
       "3     ttatccccctttcaccgtatgacgacgaatttccggtcttacaccc...  \n",
       "4     tctcactatgtggggcaggaaaagctacgcccatccgcagggtggg...  \n",
       "...                                                 ...  \n",
       "3375  atgagtcagtcagtcaatcgcgccatcgtcatcggtcacctgggcc...  \n",
       "3376  atgtgcgtcaaatcagacaacacgatccacgcgaattttacggacg...  \n",
       "3377  ttatgatgatgggtctggcagtgccggttcttccccgaaggggcga...  \n",
       "3378  tcattttcggtcagaca---ctaacgcatgagccagcatgtttcgc...  \n",
       "3379                                               None  \n",
       "\n",
       "[3380 rows x 6 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "genome_df = pd.read_csv(\n",
    "    'Genome_comparisons.csv',\n",
    "    header=None,            # Don't treat the first row as column names\n",
    "    # error_bad_lines=False,  # Skip lines with too many fields (deprecated but still functional in some versions)\n",
    "    on_bad_lines='skip',    # For newer versions of pandas\n",
    "    quoting=3,              # 3 means QUOTE_NONE: ignore quotes entirely\n",
    "    engine='python'         # Use Python engine for better error recovery\n",
    ")\n",
    "genome_df.columns = [\n",
    "    \"May_2025_Strain_Locus\", \n",
    "    \"May_2025_Nucleotide_Seq\", \n",
    "    \"Published_KRhaeticus_Locus\",\n",
    "    \"Published_KRhaeticus_Seq\",\n",
    "    \"Jan_2025_Strain_Locus\", \n",
    "    \"Jan_2025_Nucleotide_Seq\", \n",
    "]\n",
    "genome_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "99f66de7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 3380/3380 [00:00<00:00, 24887.12it/s]\n",
      "100%|██████████| 3380/3380 [00:00<00:00, 25410.27it/s]\n",
      "100%|██████████| 3380/3380 [00:00<00:00, 23239.46it/s]\n"
     ]
    }
   ],
   "source": [
    "from tqdm import tqdm\n",
    "\n",
    "def calculate_num_mutations_per_gene(col1, col2):\n",
    "    num_mutations = []\n",
    "\n",
    "    for i, row in tqdm(genome_df.iterrows(), total=len(genome_df)):\n",
    "        seq1 = row[col1]\n",
    "        seq2 = row[col2]\n",
    "\n",
    "        # Handle bad or missing data\n",
    "        if not isinstance(seq1, str) or not isinstance(seq2, str):\n",
    "            num_mutations.append(None)\n",
    "            continue\n",
    "        if len(seq1) != len(seq2):\n",
    "            num_mutations.append(None)\n",
    "            continue\n",
    "\n",
    "        matches = sum(a == b for a, b in zip(seq1, seq2))\n",
    "        # mutational edits\n",
    "        score = len(seq1) -  matches \n",
    "        num_mutations.append(score)\n",
    "    return num_mutations\n",
    "\n",
    "# Add result to DataFrame\n",
    "genome_df['Jan 2025<>May 2025 Mutations'] = calculate_num_mutations_per_gene('May_2025_Nucleotide_Seq', 'Jan_2025_Nucleotide_Seq')\n",
    "genome_df['Baseline<>May 2025 Mutations'] = calculate_num_mutations_per_gene('May_2025_Nucleotide_Seq', 'Published_KRhaeticus_Seq')\n",
    "genome_df['Baseline<>Jan 2025 Mutations'] = calculate_num_mutations_per_gene('Jan_2025_Nucleotide_Seq', 'Published_KRhaeticus_Seq')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "596d2476",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29       2.0\n",
      "63      16.0\n",
      "90       2.0\n",
      "199     14.0\n",
      "221      2.0\n",
      "        ... \n",
      "3371     2.0\n",
      "3372     6.0\n",
      "3373     2.0\n",
      "3374    50.0\n",
      "3376     6.0\n",
      "Length: 93, dtype: float64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>May_2025_Strain_Locus</th>\n",
       "      <th>May_2025_Nucleotide_Seq</th>\n",
       "      <th>Published_KRhaeticus_Locus</th>\n",
       "      <th>Published_KRhaeticus_Seq</th>\n",
       "      <th>Jan_2025_Strain_Locus</th>\n",
       "      <th>Jan_2025_Nucleotide_Seq</th>\n",
       "      <th>Jan 2025&lt;&gt;May 2025 Mutations</th>\n",
       "      <th>Baseline&lt;&gt;May 2025 Mutations</th>\n",
       "      <th>Baseline&lt;&gt;Jan 2025 Mutations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1160</th>\n",
       "      <td>0:1439595-1440314:CENAFE_06815 -</td>\n",
       "      <td>tcagagacgctcgacttccgccgccagtacatagccaccgccacgt...</td>\n",
       "      <td>1:2845765-2846484:GWK63_12985 +</td>\n",
       "      <td>tcagagacgctcgacttccgccgccagtacatagccaccgccacgt...</td>\n",
       "      <td>2:1218045-1218764:EMBJFE_01085 +</td>\n",
       "      <td>tcagagacgctcgacttccgccgccagtacatagccaccgccacgt...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2839</th>\n",
       "      <td>0:3668444-3669319:CENAFE_16895 -</td>\n",
       "      <td>tcaaccggctccacgaactgttgccatggaactggaaacccgtgaa...</td>\n",
       "      <td>1:994947-995723:GWK63_04670 +</td>\n",
       "      <td>---------------------------------------------a...</td>\n",
       "      <td>2:2565289-2566164:EMBJFE_02350 +</td>\n",
       "      <td>tcaaccggctccacgaactgttgccatggaactggaaacccgtgaa...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>45.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2238</th>\n",
       "      <td>0:2829556-2830386:CENAFE_12990 -</td>\n",
       "      <td>ttaaaaaggccgctctcccttaaccgtgacccgatacatcagccgc...</td>\n",
       "      <td>1:1792020-1792850:GWK63_08090 -</td>\n",
       "      <td>ttaaaaaggccgctctcccttaaccgtgacccgatacatcagccgc...</td>\n",
       "      <td>2:3403749-3404579:EMBJFE_03127 +</td>\n",
       "      <td>ttaaaaaggccgctctcccttaaccgtgacccgatacatcagccgc...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2781</th>\n",
       "      <td>0:3585406-3586638:CENAFE_16545 +</td>\n",
       "      <td>gtgacggcaggctgcgtgtacgttctgggcacctgcgacaccaagc...</td>\n",
       "      <td>1:1074217-1075449:GWK63_05000 -</td>\n",
       "      <td>gtgacggcaggctgcgtgtacgttctgggcacctgcgacaccaagc...</td>\n",
       "      <td>2:2648037-2649269:EMBJFE_02417 -</td>\n",
       "      <td>gtgacggcaggctgcgtgtacgttctgggcacctgcgacaccaagc...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1286</th>\n",
       "      <td>0:1586724-1587710:CENAFE_07505 -</td>\n",
       "      <td>tcatgtcgtgggttcctttccgtattgctgccagtcctcgccacac...</td>\n",
       "      <td>1:2689356-2690342:GWK63_12265 +</td>\n",
       "      <td>tcatgtcgtgggttcctttccgtattgctgccagtcctcgccacac...</td>\n",
       "      <td>2:1070649-1071635:EMBJFE_00948 +</td>\n",
       "      <td>tcatgtcgtgggttcctttccgtattgctgccagtcctcgccacac...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3009</th>\n",
       "      <td>0:1170815-1171456:CENAFE_05530 +</td>\n",
       "      <td>ttaattcggatcacggcgcttccaccgatcttcgaaccagtaaagt...</td>\n",
       "      <td>1</td>\n",
       "      <td>ttaattcggatcacggcgcttccaccgatcttcgaaccagtaaagt...</td>\n",
       "      <td>2:1485687-1486328:EMBJFE_01339 -</td>\n",
       "      <td>ttaattcggatcacggcgcttccaccgatcttcgaaccagtaaagt...</td>\n",
       "      <td>5.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>25.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3370</th>\n",
       "      <td>0</td>\n",
       "      <td>-----cctgtttgcggtcctcacgcgtcagcacgctgggccgggtc...</td>\n",
       "      <td>1:1160102-1160563:GWK63_05395 -</td>\n",
       "      <td>ttacacctgtttgcggtcctcacgcgtcagcacgctgggccgggtc...</td>\n",
       "      <td>2:3726023-3726487:EMBJFE_03445 -</td>\n",
       "      <td>ttacacctgtttgcggtcctcacgcgtcagcacgctgggccgggtc...</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>675</th>\n",
       "      <td>0:858464-861352:CENAFE_04045 -</td>\n",
       "      <td>tcaggcgctgacacccagcgccctgttcaccagcgccgcctgctcg...</td>\n",
       "      <td>1:187412-190285:GWK63_00850 +</td>\n",
       "      <td>tcaggcgctgacacccagcgccctgttcaccagcgccgcctgctcg...</td>\n",
       "      <td>2:1797543-1800416:EMBJFE_01636 +</td>\n",
       "      <td>tcaggcgctgacacccagcgccctgttcaccagcgccgcctgctcg...</td>\n",
       "      <td>15.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1410</th>\n",
       "      <td>0:1764139-1765383:CENAFE_08270 -</td>\n",
       "      <td>tcaggcgacgtcgggtacttcaatgcctgatacgaatttgaatatg...</td>\n",
       "      <td>1:2533520-2534764:GWK63_11585 +</td>\n",
       "      <td>tcaggcgacgtcgggtacttcaatgcctgatacgaatttgaatatg...</td>\n",
       "      <td>2:891660-892904:EMBJFE_00794 +</td>\n",
       "      <td>tcaggcgacgtcgggtacttcaatgcctgatacgaatttgaatatg...</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3374</th>\n",
       "      <td>0</td>\n",
       "      <td>atggaactgcgcgaagtcgatccgaaaaccctgctcgataaccccg...</td>\n",
       "      <td>1:3266634-3266864:GWK63_14995 +</td>\n",
       "      <td>atggaactgcgcgaagtcgatccgaaaaccctgctcgataaccccg...</td>\n",
       "      <td>2:3891110-3891364:EMBJFE_03606 +</td>\n",
       "      <td>atggaactgcgcgaagtcgatccgaaaaccctgctcgataaccccg...</td>\n",
       "      <td>25.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>93 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 May_2025_Strain_Locus  \\\n",
       "1160  0:1439595-1440314:CENAFE_06815 -   \n",
       "2839  0:3668444-3669319:CENAFE_16895 -   \n",
       "2238  0:2829556-2830386:CENAFE_12990 -   \n",
       "2781  0:3585406-3586638:CENAFE_16545 +   \n",
       "1286  0:1586724-1587710:CENAFE_07505 -   \n",
       "...                                ...   \n",
       "3009  0:1170815-1171456:CENAFE_05530 +   \n",
       "3370                                 0   \n",
       "675     0:858464-861352:CENAFE_04045 -   \n",
       "1410  0:1764139-1765383:CENAFE_08270 -   \n",
       "3374                                 0   \n",
       "\n",
       "                                May_2025_Nucleotide_Seq  \\\n",
       "1160  tcagagacgctcgacttccgccgccagtacatagccaccgccacgt...   \n",
       "2839  tcaaccggctccacgaactgttgccatggaactggaaacccgtgaa...   \n",
       "2238  ttaaaaaggccgctctcccttaaccgtgacccgatacatcagccgc...   \n",
       "2781  gtgacggcaggctgcgtgtacgttctgggcacctgcgacaccaagc...   \n",
       "1286  tcatgtcgtgggttcctttccgtattgctgccagtcctcgccacac...   \n",
       "...                                                 ...   \n",
       "3009  ttaattcggatcacggcgcttccaccgatcttcgaaccagtaaagt...   \n",
       "3370  -----cctgtttgcggtcctcacgcgtcagcacgctgggccgggtc...   \n",
       "675   tcaggcgctgacacccagcgccctgttcaccagcgccgcctgctcg...   \n",
       "1410  tcaggcgacgtcgggtacttcaatgcctgatacgaatttgaatatg...   \n",
       "3374  atggaactgcgcgaagtcgatccgaaaaccctgctcgataaccccg...   \n",
       "\n",
       "           Published_KRhaeticus_Locus  \\\n",
       "1160  1:2845765-2846484:GWK63_12985 +   \n",
       "2839    1:994947-995723:GWK63_04670 +   \n",
       "2238  1:1792020-1792850:GWK63_08090 -   \n",
       "2781  1:1074217-1075449:GWK63_05000 -   \n",
       "1286  1:2689356-2690342:GWK63_12265 +   \n",
       "...                               ...   \n",
       "3009                                1   \n",
       "3370  1:1160102-1160563:GWK63_05395 -   \n",
       "675     1:187412-190285:GWK63_00850 +   \n",
       "1410  1:2533520-2534764:GWK63_11585 +   \n",
       "3374  1:3266634-3266864:GWK63_14995 +   \n",
       "\n",
       "                               Published_KRhaeticus_Seq  \\\n",
       "1160  tcagagacgctcgacttccgccgccagtacatagccaccgccacgt...   \n",
       "2839  ---------------------------------------------a...   \n",
       "2238  ttaaaaaggccgctctcccttaaccgtgacccgatacatcagccgc...   \n",
       "2781  gtgacggcaggctgcgtgtacgttctgggcacctgcgacaccaagc...   \n",
       "1286  tcatgtcgtgggttcctttccgtattgctgccagtcctcgccacac...   \n",
       "...                                                 ...   \n",
       "3009  ttaattcggatcacggcgcttccaccgatcttcgaaccagtaaagt...   \n",
       "3370  ttacacctgtttgcggtcctcacgcgtcagcacgctgggccgggtc...   \n",
       "675   tcaggcgctgacacccagcgccctgttcaccagcgccgcctgctcg...   \n",
       "1410  tcaggcgacgtcgggtacttcaatgcctgatacgaatttgaatatg...   \n",
       "3374  atggaactgcgcgaagtcgatccgaaaaccctgctcgataaccccg...   \n",
       "\n",
       "                 Jan_2025_Strain_Locus  \\\n",
       "1160  2:1218045-1218764:EMBJFE_01085 +   \n",
       "2839  2:2565289-2566164:EMBJFE_02350 +   \n",
       "2238  2:3403749-3404579:EMBJFE_03127 +   \n",
       "2781  2:2648037-2649269:EMBJFE_02417 -   \n",
       "1286  2:1070649-1071635:EMBJFE_00948 +   \n",
       "...                                ...   \n",
       "3009  2:1485687-1486328:EMBJFE_01339 -   \n",
       "3370  2:3726023-3726487:EMBJFE_03445 -   \n",
       "675   2:1797543-1800416:EMBJFE_01636 +   \n",
       "1410    2:891660-892904:EMBJFE_00794 +   \n",
       "3374  2:3891110-3891364:EMBJFE_03606 +   \n",
       "\n",
       "                                Jan_2025_Nucleotide_Seq  \\\n",
       "1160  tcagagacgctcgacttccgccgccagtacatagccaccgccacgt...   \n",
       "2839  tcaaccggctccacgaactgttgccatggaactggaaacccgtgaa...   \n",
       "2238  ttaaaaaggccgctctcccttaaccgtgacccgatacatcagccgc...   \n",
       "2781  gtgacggcaggctgcgtgtacgttctgggcacctgcgacaccaagc...   \n",
       "1286  tcatgtcgtgggttcctttccgtattgctgccagtcctcgccacac...   \n",
       "...                                                 ...   \n",
       "3009  ttaattcggatcacggcgcttccaccgatcttcgaaccagtaaagt...   \n",
       "3370  ttacacctgtttgcggtcctcacgcgtcagcacgctgggccgggtc...   \n",
       "675   tcaggcgctgacacccagcgccctgttcaccagcgccgcctgctcg...   \n",
       "1410  tcaggcgacgtcgggtacttcaatgcctgatacgaatttgaatatg...   \n",
       "3374  atggaactgcgcgaagtcgatccgaaaaccctgctcgataaccccg...   \n",
       "\n",
       "      Jan 2025<>May 2025 Mutations  Baseline<>May 2025 Mutations  \\\n",
       "1160                           0.0                           1.0   \n",
       "2839                           0.0                          45.0   \n",
       "2238                           0.0                           5.0   \n",
       "2781                           0.0                           1.0   \n",
       "1286                           0.0                           1.0   \n",
       "...                            ...                           ...   \n",
       "3009                           5.0                          28.0   \n",
       "3370                           5.0                           5.0   \n",
       "675                           15.0                          20.0   \n",
       "1410                          16.0                           0.0   \n",
       "3374                          25.0                          25.0   \n",
       "\n",
       "      Baseline<>Jan 2025 Mutations  \n",
       "1160                           1.0  \n",
       "2839                          45.0  \n",
       "2238                           5.0  \n",
       "2781                           1.0  \n",
       "1286                           1.0  \n",
       "...                            ...  \n",
       "3009                          25.0  \n",
       "3370                           0.0  \n",
       "675                            5.0  \n",
       "1410                          16.0  \n",
       "3374                           0.0  \n",
       "\n",
       "[93 rows x 9 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "interesting_genes_genome_df = genome_df[(genome_df['Baseline<>Jan 2025 Mutations'] + genome_df['Baseline<>May 2025 Mutations'] + genome_df['Jan 2025<>May 2025 Mutations'] >= 1) & (genome_df['Baseline<>Jan 2025 Mutations'] + genome_df['Baseline<>May 2025 Mutations'] + genome_df['Jan 2025<>May 2025 Mutations'] < 100)]\n",
    "print(interesting_genes_genome_df[['Baseline<>Jan 2025 Mutations', 'Baseline<>May 2025 Mutations', 'Jan 2025<>May 2025 Mutations']].sum(axis=0))\n",
    "interesting_genes_genome_df.sort_values(by='Jan 2025<>May 2025 Mutations', ascending=True)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ethical_necromancy",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
